contracts/src/token/erc20/extensions/erc4626.rs

```diff
@@ -0,0 +1,525 @@
+//! ERC-4626 Tokenized Vault Standard Implementation
+//!
+//! Extends ERC-20 for vaults, enabling minting and burning of shares
+//! in exchange for underlying assets. This contract provides standardized
+//! workflows for deposits, minting, redemption, and burning of assets.
+//! Note: The shares are minted and burned in relation to the assets via
+//! the `deposit`, `mint`, `redeem`, and `burn` methods, affecting only
+//! the shares token, not the asset token itself.
+//!
+//! [ERC]: https://eips.ethereum.org/EIPS/eip-4626
+//!
+//! [CAUTION]
+//! In empty (or nearly empty) ERC-4626 vaults, deposits are at high risk of
+//! being stolen through frontrunning with a "donation" to the vault that
+//! inflates the price of a share. This is variously known as a donation or
+//! inflation attack and is essentially a problem of slippage. Vault deployers
+//! can protect against this attack by making an initial deposit of a
+//! non-trivial amount of the asset, such that price manipulation becomes
+//! infeasible. Withdrawals may similarly be affected by slippage. Users can
+//! protect against this attack as well as unexpected slippage in general by
+//! verifying the amount received is as expected, using a wrapper that performs
+//! these checks such as https://github.com/fei-protocol/ERC4626#erc4626router-and-base[ERC4626Router].
+
+use alloy_primitives::{Address, U256};
+use alloy_sol_macro::sol;
+use stylus_sdk::{
+    contract, evm, msg,
+    prelude::storage,
+    storage::{StorageU8, TopLevelStorage},
+    stylus_proc::{public, SolidityError},
+};
+
+use crate::{
+    token::erc20::{
+        self,
+        utils::{
+            safe_erc20::{self, ISafeErc20},
+            SafeErc20,
+        },
+        Erc20, IErc20,
+    },
+    utils::math::alloy::Rounding,
+};
+
+sol! {
+    /// Emitted when assets are deposited into the contract.
+    ///
+    /// * `sender` - Address of the entity initiating the deposit.
+    /// * `owner` - Address of the recipient who owns the shares.
+    /// * `assets` - Amount of assets deposited.
+    /// * `shares` - Number of shares issued to the owner.
+    #[allow(missing_docs)]
+    event Deposit(address indexed sender, address indexed owner, uint256 assets, uint256 shares);
+
+
+    /// Emitted when assets are withdrawn from the contract.
+    ///
+    /// * `sender` - Address of the entity initiating the withdrawal.
+    /// * `receiver` - Address of the recipient receiving the assets.
+    /// * `owner` - Address of the entity owning the shares.
+    /// * `assets` - Amount of assets withdrawn.
+    /// * `shares` - Number of shares burned.
+    #[allow(missing_docs)]
+    event Withdraw(
+        address indexed sender,
+        address indexed receiver,
+        address indexed owner,
+        uint256 assets,
+        uint256 shares
+    );
+}
+
+sol! {
+    /// Indicates an error where a deposit operation failed because the
```

Comment:
Adjust comments to [Solidity](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.1.0/contracts/token/ERC20/extensions/ERC4626.sol#L54-L72).